src/ice_usb_uart.c

```diff
@@ -0,0 +1,373 @@
+//
+// Created by MrJake222 on 03/11/24.
+//
+
+#include "ice_usb_uart.h"
+
+#include "rb.h"
+#include "hardware/uart.h"
+#include "hardware/irq.h"
+#include "hardware/dma.h"
+#include "hardware/timer.h"
+
+struct uart_wrap {
+    int itf;            // CDC interface number
+    uart_inst_t* inst;  // pico sdk instance reference
+
+    struct rb rx_buf;
+    struct rb tx_buf;
+
+    absolute_time_t dbg_time;
+
+    int dma_chA;
+    int dma_chB;
+    int dma_flush_alarm;
+    absolute_time_t dma_last_transfer;
+    void (*dma_flush_alarm_cb)();
+};
+
+
+// ------------------------------- Rx functions -------------------------------
+// uart -> rb (irq)
+//   rb -> cdc
+
+#define DMA_TCNT        512
+#define DMA_ALARM_US    (100*1000) // 100 ms
+
+static void ice_usb_uart_rx_dma_irq0(struct uart_wrap* uart, int channel) {
+    rb_write_ack(&uart->rx_buf, DMA_TCNT);
+
+    // complementary channel will write to write_ptr
+    // need to skip another ack
+    char* waddr = rb_write_ptr_next_ack(&uart->rx_buf, DMA_TCNT);
+    dma_channel_set_write_addr(channel, waddr, false);
```

Comment:
the interleaving is happening at a hardware level (chain_to field), This is done to avoid reconfiguring the channels when there is new data to be processed. When chA is being reconfigured, chB already copies data to it's portion of the buffer.